support/zlib-helper.c

```diff
@@ -208,7 +208,9 @@ ReadZStream (ZStream *stream, guchar *buffer, gint length)
 			stream->eof = TRUE;
 			break;
 		} else if (status == Z_BUF_ERROR && stream->total_in == zs->total_in) {
-			stream->eof = TRUE;
+			if (zs->avail_in != 0) {
```

Comment:
I am not sure this is this right place. Could we skip everything when the count is 0 at https://github.com/mono/mono/blob/198150688b7087450ba67df2e2437dd747bab5aa/support/zlib-helper.c#L188 ?

Reply:
We cannot and there are unit tests for it. There could be partially decoded bytes from previous call that didn't fit in the output buffer. Thus even reading 0 bytes from input stream can produce some output.

Reply:
To further elaborate: Lot of the code, including `stream->eof` and https://github.com/mono/mono/blob/198150688b7087450ba67df2e2437dd747bab5aa/support/zlib-helper.c#L203-204 are just unnecessary shortcuts and micro-optimizations. This was the least invasive place I could find to solve the issue.